controllers/uploadpackage.lua

```diff
@@ -0,0 +1,108 @@
+local app_helpers = require("lapis.application")
+local validate = require("lapis.validate")
+local capture_errors, assert_error = app_helpers.capture_errors, app_helpers.assert_error
+local lfs = require('lfs')
+local Packages = require("models.packages")
+local Users = require("models.users")
+
+local function isFile(name)
+  if not isDir(name) then
+    return os.rename(name,name) and true or false
+  end
+  return false
+end
+
+local function isFileOrDir(name)
+  if type(name)~="string" then return false end
+  return os.rename(name, name)
+end
+
+local function isDir(name)
+  if type(name)~="string" then return false end
+  local cd = lfs.currentdir()
+  local is, err = lfs.chdir(name)
+  lfs.chdir(cd)
+  return is, err
+end
+
+return {
+  GET = capture_errors(function(self)
+    assert_error(self.session.name, self.i18n("err_not_logged_in"))
+    assert_error(self.session.verified, self.i18n("err_email_not_verified"))
+    return { render = "uploadpackage" }
+  end),
+  POST = capture_errors(function(self)
+    -- must be logged in to upload
+    assert_error(self.session.name, self.i18n("err_not_logged_in"))
+    
+    -- must provide a file, a name, and a version
+    validate.assert_valid(self.params, {
+      { "name", exists = true, min_length = 2 },
+      { "version", exists = true },
+      { "file", exists = true, is_file = true }
+    })
+    -- prep some info
+    local fname = self.params.file.filename
+    local fcontent = self.params.file.content
+    local zipheader = "PK"
+    local mudlet_xml_header = "<!DOCTYPE MudletPackage>"
+    local file_extension = self.split_string(fname, "%.")[2]
+    
+    -- validate file extension
+    local proper_extension = file_extension and (file_extension == "mpackage" or file_extension == "xml")
```

Comment:
`.zip` is also a valid extension, same behaviour as `.mpackage`.

Reply:
Can do this, but I'm hesitant to open up to just .zip files. With the email confirmation and requiring login to do it I'm slightly less hesitant, but I'd hate to find out some skript kiddie sussed it out and was using us as a distro site.

Reply:
They still could... I think user convenience is important to aim for. Let's try it and see if it bites too much?